pkg/generators/matrix.go

```diff
@@ -0,0 +1,82 @@
+package generators
+
+import (
+	"errors"
+	"time"
+
+	argoprojiov1alpha1 "github.com/argoproj-labs/applicationset/api/v1alpha1"
+	"github.com/argoproj-labs/applicationset/pkg/utils"
+)
+
+var _ Generator = (*MatrixGenerator)(nil)
+
+var MoreThanTwoGenerators = errors.New("found more than two generators, Matrix support only two")
+var LessThanTwoGenerators = errors.New("found less than two generators, Matrix support only two")
+var MoreThenOneInnerGenerators = errors.New("found more than one generator in matrix.Generators")
+
+type MatrixGenerator struct {
+	generators map[string]Generator
+}
+
+func NewMatrixGenerator(generators map[string]Generator) Generator {
+	m := &MatrixGenerator{
+		generators: generators,
+	}
+	return m
+}
+
+func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) ([]map[string]string, error) {
+
+	if len(appSetGenerator.Matrix.Generators) < 2 {
+		return nil, LessThanTwoGenerators
+	}
+
+	if len(appSetGenerator.Matrix.Generators) > 2 {
+		return nil, MoreThanTwoGenerators
+	}
+
+	res := []map[string]string{}
+
+	g0, err := m.getParams(appSetGenerator.Matrix.Generators[0])
+	if err != nil {
+		return nil, err
+	}
+	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1])
+	if err != nil {
+		return nil, err
+	}
+
+	for _, a := range g0 {
+		for _, b := range g1 {
+			res = append(res, utils.CombineStringMaps(a, b))
+		}
+	}
+
+	return res, nil
+}
+
+func (m *MatrixGenerator) getParams(appSetBaseGenerator argoprojiov1alpha1.ApplicationSetBaseGenerator) ([]map[string]string, error) {
+
+	t, _ := Transform(
+		argoprojiov1alpha1.ApplicationSetGenerator{
+			List:     appSetBaseGenerator.List,
+			Clusters: appSetBaseGenerator.Clusters,
+			Git:      appSetBaseGenerator.Git,
+		},
+		m.generators,
+		argoprojiov1alpha1.ApplicationSetTemplate{})
+
+	if len(t) > 1 {
+		return nil, MoreThenOneInnerGenerators
+	}
+
+	return t[0].Params, nil
+}
+
+func (m *MatrixGenerator) GetRequeueAfter(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) time.Duration {
+	return NoRequeueAfter
```

Comment:
I think for Matrix generator's `GetRequeueAfter`, it should call `GetRequeueAfter` on all the inner generators, and returned the largest value received? Otherwise we might be suppressing a non-zero RequeueAfter value from an inner generator (unless this is handled elsewhere).